File: news-social-intelligence-data-pipeline/social_intel/youtube.py
"""YouTube metadata scraper using yt-dlp (no API key required).

Uses yt-dlp binary to extract video metadata without downloading.
Requires: pip install yt-dlp (or system package)

Output format matches our normalized article schema.
"""

import json
import subprocess
import sys
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from .lib import combined_relevance

YTDLP_TIMEOUT = 30

# Depth-aware limits
DEPTH_LIMITS = {
    "quick": 5,
    "default": 10,
    "deep": 25,
}


def _log(msg: str) -> None:
    sys.stderr.write(f"[YouTube] {msg}\n")
    sys.stderr.flush()


def _run_ytdlp(args: List[str], timeout: int = YTDLP_TIMEOUT) -> Optional[str]:
    """Run yt-dlp command and return stdout. Never raises."""
    cmd = ["yt-dlp", "--no-warnings", "-q"] + args
    
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        if result.returncode == 0:
            return result.stdout
        else:
            _log(f"yt-dlp error: {result.stderr[:200]}")
            return None
    except subprocess.TimeoutExpired:
        _log(f"yt-dlp timeout after {timeout}s")
        return None
    except FileNotFoundError:
        _log("yt-dlp not found - install with: pip install yt-dlp")
        return None
    except Exception as e:
        _log(f"yt-dlp error: {e}")
        return None


def _parse_video(data: Dict[str, Any], query: str = "") -> Optional[Dict[str, Any]]:
    """Parse yt-dlp JSON into normalized article dict. Never raises."""
    try:
        video_id = data.get("id", "")
        title = data.get("title", "")
        
        if not video_id or not title:
            return None
        
        description = data.get("description", "") or ""
        description = description[:500]
        
        # Calculate relevance
        relevance = combined_relevance(query, title, description) if query else 0.0
        
        # Parse upload date (YYYYMMDD format)
        upload_date = data.get("upload_date")
        date_str = None
        if upload_date and len(upload_date) == 8:
            try:
                date_str = f"{upload_date[:4]}-{upload_date[4:6]}-{upload_date[6:8]}"
            except Exception:
                pass
        
        return {
            "id": video_id,
            "title": title,
            "url": data.get("webpage_url") or f"https://www.youtube.com/watch?v={video_id}",
            "description": description,
            "source": "YouTube",
            "author": data.get("uploader") or data.get("channel") or "",
            "score": data.get("view_count") or 0,
            "num_comments": data.get("comment_count") or 0,
            "created_utc": data.get("timestamp"),
            "date": date_str,
            "relevance": round(relevance, 3),
            "why_relevant": "yt-dlp search",
            "collected_at": datetime.now(timezone.utc).isoformat(),
            "content_type": "youtube",
            "metadata": {
                "channel_id": data.get("channel_id"),
                "channel": data.get("channel"),
                "duration": data.get("duration"),
                "view_count": data.get("view_count"),
                "like_count": data.get("like_count"),
                "categories": data.get("categories", []),
                "tags": data.get("tags", [])[:10],
            },
        }
    except Exception as e:
        _log(f"Parse error: {e}")
        return None


def search_youtube(
    query: str,
    depth: str = "default",
) -> List[Dict[str, Any]]:
    """Search YouTube videos. Never raises.
    
    Args:
        query: Search query string
        depth: 'quick', 'default', or 'deep' - controls result limit
        
    Returns:
        List of normalized article dicts, empty on failure
    """
    limit = DEPTH_LIMITS.get(depth, DEPTH_LIMITS["default"])
    
    _log(f"Searching: {query} (depth={depth}, limit={limit})")
    
    # yt-dlp search syntax
    search_url = f"ytsearch{limit}:{query}"
    
    output = _run_ytdlp([
        "--dump-json",
        "--flat-playlist",
        search_url,
    ])
    
    if not output:
        return []
    
    results = []
    for line in output.strip().split("\n"):
        if not line:
            continue
        try:
            data = json.loads(line)
            parsed = _parse_video(data, query)
            if parsed:
                results.append(parsed)
        except json.JSONDecodeError:
            continue
    
    _log(f"Found {len(results)} results")
    return results


def get_video_info(url: str) -> Optional[Dict[str, Any]]:
    """Get metadata for a specific video. Never raises.
    
    Args:
        url: YouTube video URL
        
    Returns:
        Normalized article dict or None on failure
    """
    _log(f"Getting info: {url}")
    
    output = _run_ytdlp([
        "--dump-json",
        "--no-playlist",
        url,
    ])
    
    if not output:
        return None
    
    try:
        data = json.loads(output)
        return _parse_video(data)
    except json.JSONDecodeError:
        return None


def get_channel_videos(
    channel_url: str,
    limit: int = 10,
) -> List[Dict[str, Any]]:
    """Get recent videos from a channel. Never raises.
    
    Args:
        channel_url: YouTube channel URL (/@handle or /channel/ID)
        limit: Max videos to return
        
    Returns:
        List of normalized article dicts
    """
    _log(f"Getting channel videos: {channel_url}")
    
    # Add /videos to get uploads
    if not channel_url.endswith("/videos"):
        channel_url = channel_url.rstrip("/") + "/videos"
    
    output = _run_ytdlp([
        "--dump-json",
        "--flat-playlist",
        "--playlist-end", str(limit),
        channel_url,
    ], timeout=60)  # Channels can take longer
    
    if not output:
        return []
    
    results = []
    for line in output.strip().split("\n"):
        if not line:
            continue
        try:
            data = json.loads(line)
            parsed = _parse_video(data)
            if parsed:
                results.append(parsed)
        except json.JSONDecodeError:
            continue
    
    _log(f"Found {len(results)} videos")
    return results


if __name__ == "__main__":
    print("=== Search: Python tutorial ===")
    results = search_youtube("Python tutorial", depth="quick")
    for item in results:
        views = item["metadata"].get("view_count", 0)
        print(f"  [{views:,} views] {item['title'][:50]}")

File: news-social-intelligence-data-pipeline/src/social/youtube.rs
//! YouTube collector via yt-dlp subprocess (no API key required).

use serde::Deserialize;
use std::process::Command;
use tracing::{info, warn};
use chrono::Utc;

use super::SocialArticle;
use super::relevance::combined_relevance;

/// Timeout for yt-dlp subprocess calls (reserved for future use with process timeout).
#[allow(dead_code)]
const YTDLP_TIMEOUT_SECS: u64 = 30;

/// Parsed yt-dlp JSON output for a single video.
#[derive(Debug, Deserialize)]
struct YtDlpOutput {
    id: Option<String>,
    title: Option<String>,
    description: Option<String>,
    webpage_url: Option<String>,
    uploader: Option<String>,
    channel: Option<String>,
    channel_id: Option<String>,
    upload_date: Option<String>, // "YYYYMMDD"
    timestamp: Option<i64>,
    view_count: Option<i64>,
    like_count: Option<i64>,
    comment_count: Option<i64>,
    duration: Option<i64>,
    categories: Option<Vec<String>>,
    tags: Option<Vec<String>>,
}

/// Run yt-dlp with the given arguments and return stdout on success.
/// Returns `None` on any error (binary not found, timeout, non-zero exit).
fn run_ytdlp(args: &[&str]) -> Option<String> {
    let mut cmd_args = vec!["--no-warnings", "-q"];
    cmd_args.extend_from_slice(args);

    let output = match Command::new("yt-dlp")
        .args(&cmd_args)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()
    {
        Ok(o) => o,
        Err(e) => {
            warn!("yt-dlp not found or failed to spawn: {}", e);
            return None;
        }
    };

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        warn!("yt-dlp exited with status {}: {}", output.status, &stderr[..stderr.len().min(200)]);
        return None;
    }

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    if stdout.is_empty() {
        None
    } else {
        Some(stdout)
    }
}

/// Convert a YtDlpOutput into a normalized SocialArticle.
/// Returns `None` if required fields (id, title) are missing.
fn parse_video(data: &YtDlpOutput, query: &str) -> Option<SocialArticle> {
    let video_id = data.id.as_deref()?;
    let title = data.title.as_deref()?;

    if video_id.is_empty() || title.is_empty() {
        return None;
    }

    let description = data
        .description
        .as_deref()
        .unwrap_or("")
        .chars()
        .take(500)
        .collect::<String>();

    let relevance = if query.is_empty() {
        0.0
    } else {
        combined_relevance(query, title, &description, 0.7)
    };

    // Parse upload_date "YYYYMMDD" → "YYYY-MM-DD"
    let date = data.upload_date.as_deref().and_then(|d| {
        if d.len() == 8 {
            Some(format!("{}-{}-{}", &d[..4], &d[4..6], &d[6..8]))
        } else {
            None
        }
    });

    let url = data
        .webpage_url
        .clone()
        .unwrap_or_else(|| format!("https://www.youtube.com/watch?v={}", video_id));

    let author = data
        .uploader
        .as_deref()
        .or(data.channel.as_deref())
        .unwrap_or("")
        .to_string();

    let view_count = data.view_count.unwrap_or(0);
    let comment_count = data.comment_count.unwrap_or(0);

    let tags: Vec<String> = data
        .tags
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .take(10)
        .cloned()
        .collect();

    let metadata = serde_json::json!({
        "channel_id": data.channel_id,
        "channel": data.channel,
        "duration": data.duration,
        "view_count": data.view_count,
        "like_count": data.like_count,
        "categories": data.categories.as_deref().unwrap_or(&[]),
        "tags": tags,
    });

    Some(SocialArticle {
        id: video_id.to_string(),
        title: title.to_string(),
        url,
        description,
        source: "YouTube".to_string(),
        author,
        score: view_count,
        num_comments: comment_count,
        created_utc: data.timestamp.map(|t| t as f64),
        date,
        relevance,
        collected_at: Utc::now().to_rfc3339(),
        content_type: "youtube".to_string(),
        metadata,
    })
}

/// Search YouTube for videos matching the query.
/// Uses `yt-dlp --flat-playlist` to get metadata without downloading.
/// Returns an empty Vec on any error.
pub fn search_youtube(query: &str, limit: usize) -> Vec<SocialArticle> {
    let search_url = format!("ytsearch{}:{}", limit, query);
    info!("🎬 YouTube search: {} (limit={})", query, limit);

    let output = match run_ytdlp(&["--dump-json", "--flat-playlist", &search_url]) {
        Some(o) => o,
        None => return Vec::new(),
    };

    let mut results = Vec::new();
    for line in output.lines() {
        if line.trim().is_empty() {
            continue;
        }
        match serde_json::from_str::<YtDlpOutput>(line) {
            Ok(data) => {
                if let Some(article) = parse_video(&data, query) {
                    results.push(article);
                }
            }
            Err(e) => {
                warn!("Failed to parse yt-dlp JSON line: {}", e);
            }
        }
    }

    info!("🎬 YouTube found {} results", results.len());
    results
}

/// Get metadata for a specific YouTube video URL.
/// Returns `None` on any error.
pub fn get_video_info(url: &str) -> Option<SocialArticle> {
    info!("🎬 YouTube get info: {}", url);

    let output = run_ytdlp(&["--dump-json", "--no-playlist", url])?;

    let data: YtDlpOutput = serde_json::from_str(&output).ok()?;
    parse_video(&data, "")
}

/// Get recent videos from a YouTube channel.
/// Returns an empty Vec on any error.
pub fn get_channel_videos(channel_url: &str, limit: usize) -> Vec<SocialArticle> {
    info!("🎬 YouTube channel videos: {} (limit={})", channel_url, limit);

    let url = if channel_url.ends_with("/videos") {
        channel_url.to_string()
    } else {
        format!("{}/videos", channel_url.trim_end_matches('/'))
    };

    let limit_str = limit.to_string();
    let output = match run_ytdlp(&[
        "--dump-json",
        "--flat-playlist",
        "--playlist-end",
        &limit_str,
        &url,
    ]) {
        Some(o) => o,
        None => return Vec::new(),
    };

    let mut results = Vec::new();
    for line in output.lines() {
        if line.trim().is_empty() {
            continue;
        }
        match serde_json::from_str::<YtDlpOutput>(line) {
            Ok(data) => {
                if let Some(article) = parse_video(&data, "") {
                    results.push(article);
                }
            }
            Err(e) => {
                warn!("Failed to parse yt-dlp JSON line: {}", e);
            }
        }
    }

    info!("🎬 YouTube channel found {} videos", results.len());
    results
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_ytdlp_output() -> YtDlpOutput {
        YtDlpOutput {
            id: Some("dQw4w9WgXcQ".to_string()),
            title: Some("Rick Astley - Never Gonna Give You Up".to_string()),
            description: Some("The official video for Rick Astley".to_string()),
            webpage_url: Some("https://www.youtube.com/watch?v=dQw4w9WgXcQ".to_string()),
            uploader: Some("Rick Astley".to_string()),
            channel: Some("Rick Astley".to_string()),
            channel_id: Some("UCuAXFkgsw1L7xaCfnd5JJOw".to_string()),
            upload_date: Some("20091025".to_string()),
            timestamp: Some(1256428800),
            view_count: Some(1_500_000_000),
            like_count: Some(15_000_000),
            comment_count: Some(3_000_000),
            duration: Some(212),
            categories: Some(vec!["Music".to_string()]),
            tags: Some(vec![
                "rick astley".to_string(),
                "never gonna give you up".to_string(),
                "music".to_string(),
            ]),
        }
    }

    #[test]
    fn test_parse_video_valid() {
        let data = sample_ytdlp_output();
        let article = parse_video(&data, "rick astley").unwrap();

        assert_eq!(article.id, "dQw4w9WgXcQ");
        assert_eq!(article.title, "Rick Astley - Never Gonna Give You Up");
        assert_eq!(article.url, "https://www.youtube.com/watch?v=dQw4w9WgXcQ");
        assert_eq!(article.source, "YouTube");
        assert_eq!(article.content_type, "youtube");
        assert_eq!(article.author, "Rick Astley");
        assert_eq!(article.score, 1_500_000_000);
        assert_eq!(article.num_comments, 3_000_000);
        assert_eq!(article.date, Some("2009-10-25".to_string()));
        assert_eq!(article.created_utc, Some(1256428800.0));
        assert!(article.relevance > 0.0);

        // Check metadata
        let meta = &article.metadata;
        assert_eq!(meta["channel_id"], "UCuAXFkgsw1L7xaCfnd5JJOw");
        assert_eq!(meta["channel"], "Rick Astley");
        assert_eq!(meta["duration"], 212);
        assert_eq!(meta["view_count"], 1_500_000_000i64);
        assert_eq!(meta["like_count"], 15_000_000);
    }

    #[test]
    fn test_parse_video_missing_title() {
        let data = YtDlpOutput {
            id: Some("abc123".to_string()),
            title: None,
            description: None,
            webpage_url: None,
            uploader: None,
            channel: None,
            channel_id: None,
            upload_date: None,
            timestamp: None,
            view_count: None,
            like_count: None,
            comment_count: None,
            duration: None,
            categories: None,
            tags: None,
        };

        let result = parse_video(&data, "test");
        assert!(result.is_none());
    }

    #[test]
    fn test_upload_date_conversion() {
        let data = YtDlpOutput {
            id: Some("test123".to_string()),
            title: Some("Test Video".to_string()),
            description: None,
            webpage_url: None,
            uploader: None,
            channel: None,
            channel_id: None,
            upload_date: Some("20260610".to_string()),
            timestamp: None,
            view_count: None,
            like_count: None,
            comment_count: None,
            duration: None,
            categories: None,
            tags: None,
        };

        let article = parse_video(&data, "").unwrap();
        assert_eq!(article.date, Some("2026-06-10".to_string()));
    }

    #[test]
    fn test_run_ytdlp_not_found() {
        // run_ytdlp uses "yt-dlp" binary. If not installed, returns None.
        // On machines with yt-dlp, this will succeed but we just verify no panic.
        let result = run_ytdlp(&["--version"]);
        // Either Some (if installed) or None (if not) — both valid.
        let _ = result;
    }
}
